ddclient.in

```diff
@@ -2020,114 +2022,176 @@ sub geturl {
     debug("url      = %s", $url);
     debug("ip ver   = %s", $ipversion);
 
-    ## determine peer and port to use.
-    $peer = $proxy || $server;
-    $peer =~ s%/.*%%;
-    $port = $peer;
-    $port =~ s%^.*:%%;
-    $port = $default_port unless $port =~ /^\d+$/;
-    $peer =~ s%:.*$%%;
-
-    $request = "$method ";
-    if (!$use_ssl) {
-        $request .= "http://$server" if $proxy;
-    } else {
-        $request .= "https://$server" if $proxy;
-    }
-    $request .= "/$url HTTP/1.0\n";
-    $request .= "Host: $server\n";
-
-    my $auth = encode_base64("${login}:${password}", "");
-    $request .= "Authorization: Basic $auth\n" if $login || $password;
-    $request .= "User-Agent: ${program}/${version}\n";
-    if ($data) {
-        $request .= "Content-Type: application/x-www-form-urlencoded\n" if $headers !~ /^Content-Type:/mi;
-        $request .= "Content-Length: " . length($data) . "\n";
-    }
-    $request .= "Connection: close\n";
-    $headers .= "\n" if $headers ne '' && substr($headers, -1) ne "\n";
-    $request .= $headers;
-    $request .= "\n";
-    # RFC 7230 says that all lines before the body must end with <cr><lf>.
-    (my $rq = $request) =~ s/(?<!\r)\n/\r\n/g;
-    $request .= $data;
-    $rq .= $data;
-
-    my %socket_args = (
-        PeerAddr => $peer,
-        PeerPort => $port,
-        Proto => 'tcp',
-        MultiHomed => 1,
-        Timeout => opt('timeout'),
-    );
-    my $socket_class = 'IO::Socket::INET';
-    if ($use_ssl) {
-        # IO::Socket::SSL will load IPv6 support if available on the system.
-        load_ssl_support;
-        $socket_class = 'IO::Socket::SSL';
-        $socket_args{SSL_ca_file} = opt('ssl_ca_file') if defined(opt('ssl_ca_file'));
-        $socket_args{SSL_ca_path} = opt('ssl_ca_dir') if defined(opt('ssl_ca_dir'));
-    } elsif ($globals{'ipv6'} || $ipversion eq '6') {
-        load_ipv6_support;
-        $socket_class = 'IO::Socket::INET6';
-    }
-    if ($ipversion eq '4') {
-        $socket_args{Domain} = PF_INET;
-        $socket_args{Family} = AF_INET;
-    } elsif ($ipversion eq '6') {
-        $socket_args{Domain} = PF_INET6;
-        $socket_args{Family} = AF_INET6;
-    } elsif ($ipversion ne '') {
-        fatal("geturl passed unsupported 'ipversion' value %s", $ipversion);
-    }
+    if (opt('curl')) {
```

Comment:
Please split this one function into three to make the code easier to read and test, and to make this PR easier to review:

```perl
sub geturl {
    # ...
    my @args = (...);
    my $reply = opt('curl') ? fetch_via_curl(@args) : fetch_via_socket_io(@args);
    # ...
}
```

Reply:
Done.